docs/proposals/webhook-validation.md

```diff
@@ -0,0 +1,130 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+Do we want to require that operators run another job to synchronously validate client requests?
+
+Do we want to rely on Knative's webhook packages, or write our own code to manage certs ourselves?
+Is there anything already out there in controller-gen we should be using?
+
+## Summary
+
+Kubernetes provides facilities to inject synchronous validation* of CRUD operations from clients for specified types, such as Shipwright's CRD types.
+
+_*and mutation, including default-setting, but these are not described in this proposal._
+
+We should take advantage of these features and include a webhook validation deployment in the Shipwright installation to provide this functionality.
+
+## Motivation
+
+Today, when a user creates a resource (e.g., a Build, BuildRun, etc.), they are immediately stored in the cluster's etcd storage, and the Shipwright controller begins reconciling the object.
+
+If the controller finds that the resource is invalid, it updates the resource's `.status` to indicate this state and stops reconciling it.
+
+Examples of invalid resources include:
+
+- a BuildRun referencing a Build that doesn't exist
+- a BuildRun referencing a ServiceAccount that doesn't exist
+- a Build referencing a (Cluster)BuildStrategy that doesn't exist
+- a BuildStrategy containing zero steps
+
+(This is a partial list, and new validation can be assumed to be added over time as the Shipwright API evolves.)
+
+By allowing invalid resources to be accepted and stored, we require clients to watch or poll all resources to ensure their request was valid, and we add to the API surface of resources that otherwise would not need to report a status, like a Build.
+
+Always accepting and storing resources, even invalid ones, can also lead to additional storage of known-invalid objects, which clutter the user experience and can lead to confusion, and in extreme cases, cluster resource exhaustion.
+
+Mixing reconciliation and validation code in the controller also complicates our codebase, leading to PRs like https://github.com/shipwright-io/build/pull/641 which add new validation-centric statuses that the user must watch for.
+As new validation scenarios are added, so too will the number of meaningful status strings, which must be documented, tested, etc.
+
+### Goals
+
+Enforce synchronous validation via a new Deployment and a [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook), and remove validation-centric logic from the reconciler.
+
+The reconciler should be able to assume in many cases that the resource it is reconciling has already been validated.
+
+### Non-Goals
+
+- Include a mutating or defaulting or conversion webhook.
+  If we decide these are useful, we can add them in future proposals.
```

Comment:
Such as the default value for timeout or the default value for the BuildStrategy type, I think if we decide to add webhook, these defaulting are also useful for us, but can be added in future.

Reply:
Yep, those are definitely things we should investigate after we have some webhook in place.